stablehlo/dialect/TypeInference.cpp

```diff
@@ -2046,7 +2047,13 @@ LogicalResult inferDynamicGatherOp(
                           collapsedSliceDims, startIndexMap, indexVectorDim)))
     return failure();
 
-  auto getSliceDim = [](int64_t index) { return ShapedType::kDynamic; };
+  auto getSliceDim = [&](int64_t index) {
+    DenseIntElementsAttr sliceSizesAttr;
+    if (!matchPattern(sliceSizes, m_Constant(&sliceSizesAttr))) {
+      return ShapedType::kDynamic;
+    }
+    return sliceSizesAttr.getValues<APInt>()[index].getSExtValue();
```

Comment:
LGTM but with this comment:
However, now we have the power of use `getSExtValue()` for a Value if it's constant. Is there a plan to apply this refinement to many op, or, for this gather op only?

Reply:
Good question! In principle, this idea is applicable to all ops whose result shapes are computed based on operands: `dynamic_broadcast_in_dim`, `dynamic_conv`, `dynamic_gather`, `dynamic_iota`, `dynamic_pad`, `dynamic_reshape`, `real_dynamic_slice`, `rng`.

However, in practice we may want to postpone implementing this idea until the Python API is figured out, as discussed in https://github.com/openxla/stablehlo/issues/863#issuecomment-1407562475. I'm even thinking to propose removing the shape function for DynamicGatherOp in the meanwhile, although specifically for this op it doesn't matter much logistically, because JAX is already using a workaround.